packages/ember-data/lib/system/snapshot.js

```diff
@@ -0,0 +1,298 @@
+        // causes an infinite loop if accessed and calls `Ember.deprecate()`
+        // again.
+        if (callDeprecate) {
+          callDeprecate = false;
+          Ember.deprecate('Usage of `snapshot.constructor` is deprecated, use `snapshot.type` instead.');
+          callDeprecate = true;
+        }
+
+        return this.type;
+      }
+    });
+  } else {
+    this.constructor = this.type;
+  }
+}
+
+Snapshot.prototype = {
+  constructor: Snapshot,
+
+  /**
+    Returns the value of an attribute.
+
+    Example
+
+    ```javascript
+    var post = store.createRecord('post', { author: 'Tomster', title: 'Ember.js rocks' });
+    var snapshot = post._createSnapshot();
+
+    snapshot.attr('author'); // => 'Tomster'
+    snapshot.attr('title'); // => 'Ember.js rocks'
+    ```
+
+    Note: Values are loaded eagerly and cached when the snapshot is created.
+
+    @method attr
+    @param {string} keyName
+    @return {Object} The attribute value or undefined
+  */
+  attr: function(keyName) {
+    if (keyName in this._attributes) {
+      return this._attributes[keyName];
+    }
+    throw new Ember.Error("Model '" + Ember.inspect(this.record) + "' has no attribute named '" + keyName + "' defined.");
+  },
+
+  /**
+    Returns all attributes and their corresponding values.
+
+    Example
+
+    ```javascript
+    var post = store.createRecord('post', { author: 'Tomster', title: 'Ember.js rocks' });
+    var snapshot = post._createSnapshot();
+
+    snapshot.attributes(); // => { author: 'Tomster', title: 'Ember.js rocks' }
+    ```
+
+    @method attributes
+    @return {Array} All attributes for the current snapshot
+  */
+  attributes: function() {
+    return Ember.copy(this._attributes);
+  },
+
+  /**
+    Returns the id of the snapshot's record
+
+    Example
+
+    ```javascript
+    var post = store.push('post', { id: 1, author: 'Tomster', title: 'Ember.js rocks' });
+    var snapshot = post._createSnapshot();
+
+    snapshot.id(); // => '1'
+    ```
+
+    @method id
+    @return {string} The ID of the snapshot's record
+  */
+  id: function() {
+    return this._id;
+  },
+
+  /**
+    Returns the current value of a belongsTo relationship.
+
+    Example
+
+    ```javascript
+    var post = store.createRecord('post', { title: 'Hello World' });
+    var comment = store.createRecord('comment', { body: 'Lorem ipsum', post: post });
+    var snapshot = comment._createSnapshot();
+
+    snapshot.belongsTo('post'); // => DS.Snapshot of post
+    ```
+
+    Calling `belongsTo` will return a new Snapshot as long as there's any
+    data available, such as an ID. If there's no data available `belongsTo` will
+    return undefined.
+
+    Note: Relationships are loaded lazily and cached upon first access.
+
+    @method belongsTo
+    @param {string} keyName
+    @return {DS.Snapshot|undefined} A snapshot of a belongsTo relationship or undefined
+  */
+  belongsTo: function(keyName) {
```

Comment:
Yeah, that might be useful. There's one place that could benefit from this, see https://github.com/emberjs/data/pull/2623/files#diff-874af7e7df6c1be12e3a54caad8d859dR583

However, `ids` feels wrong to me since it's just one id? But `id` would be inconsistent to `hasMany`.

@igorT ?